host/xtest/sdp_basic.c

```diff
@@ -0,0 +1,555 @@
+#define RANDOM_BUFFER_SIZE	(32 * 1024 + 1)
+static int get_random_bytes(char *out, size_t len)
+{
+	static char *rand_buf = NULL;
+	static size_t rand_idx = 0;
+	int rc;
+
+	if (!rand_buf) {
+		const char rand_dev[] = "/dev/urandom";
+		int fd;
+
+		rand_buf = malloc(RANDOM_BUFFER_SIZE);
+		if (!rand_buf) {
+			fprintf(stderr, "failed to random buffer memory (%d bytes)\n",
+				RANDOM_BUFFER_SIZE);
+			return -1;
+		}
+
+		fd = open(rand_dev, O_RDONLY);
+		if (fd < 0) {
+			fprintf(stderr, "failed to open %s\n", rand_dev);
+			return -1;
+		}
+
+		rc = read(fd, rand_buf, RANDOM_BUFFER_SIZE);
+		if (rc != RANDOM_BUFFER_SIZE) {
+			fprintf(stderr, "failed to read %d bytes from %s\n",
+				RANDOM_BUFFER_SIZE, rand_dev);
+			return -1;
+		}
+		close(fd);
+	}
+
+	while (len) {
+		size_t t_len = (RANDOM_BUFFER_SIZE < len) ? RANDOM_BUFFER_SIZE : len;
+
+		if ((rand_idx + t_len) > RANDOM_BUFFER_SIZE) {
+			int sz_end = RANDOM_BUFFER_SIZE - rand_idx;
+			int sz_beg = t_len - sz_end;
+
+			memcpy(out, rand_buf + rand_idx, sz_end);
+			memcpy(out + sz_end, rand_buf , sz_beg);
+			rand_idx = sz_beg;
+		} else {
+			memcpy(out, rand_buf + rand_idx, t_len);
+			rand_idx += t_len;
+		}
+		len -= t_len;
+	}
+	return 0;
+}
+
+
+static int sdp_basic_test(size_t size, size_t loop, int ion_heap, int ns_map)
+{
+	struct tee_ctx *ctx = NULL;
+	unsigned char *test_buf = NULL;
+	unsigned char *ref_buf = NULL;
+	void *shm_ref = NULL;
+	unsigned int err = 1;
+	int fd = -1;
+	size_t offset;
+
+	if (!loop) {
+		fprintf(stderr, "Error: null loop value\n");
+		return 1;
+	}
+
+	test_buf = malloc(size);
+	ref_buf = malloc(size);
+	if (!test_buf || !ref_buf) {
+		verbose("failed to allocate memory\n");
+		goto out;
+	}
+
+	/* create secure buffer (allocate 'size' + max 'offset' (255) */
+	fd = allocate_ion_buffer(size + 1024, ion_heap);
+	if (fd < 0) {
+		verbose("Failed to allocate SDP buffer (%d bytes) in ION heap %d: %d\n",
+				size + 1024, ion_heap, fd);
+		goto out;
+	}
+	if (ns_map) {
+		char *d = mmap(NULL, size + 1024, PROT_READ, MAP_SHARED, fd, 0);
+
+		if (d == MAP_FAILED) {
+			fprintf(stderr, "Note: cannot map/check SDP buffers\n");
+			ns_map = 0;
+		} else
+			munmap(d, size + 1024);
+	}
+
+
+	/* register secure buffer to TEE */
+	ctx = malloc(sizeof(*ctx));
+	if (!ctx)
+		goto out;
+	if (create_tee_ctx(ctx))
+		goto out;
+	if (tee_register_buffer(ctx, &shm_ref, fd))
+		goto out;
+
+	/* release registered fd: tee should still hold refcount on resource */
+	close(fd);
+	fd = -1;
+
+	/* invoke trusted application with secure buffer as memref parameter */
+	while (loop--) {
+		/* get an buffer of random-like values */
+		if (get_random_bytes((char *)ref_buf, size))
+			goto out;
+		memcpy(test_buf, ref_buf, size);
+		/* random offset [0 255] */
+		offset = (unsigned int)*ref_buf;
+
+		/* TA writes into SDP buffer */
+		if (inject_sdp_data(ctx, test_buf, offset, size, shm_ref))
+			goto out;
+
+		/* attempt to map check SDP buffer content */
+		if (ns_map &&
```

Comment:
the test is:
#1- invoke the TA to fill the secure buffer with data from a nonsecure (shm) buffer.
#2- invoke the TA to process the secure buffer.
#3- invoke the TA to dump the secure buffer into a nonsecure (shm) buffer.
  (so we can check dumped data have the expected values)

After stage #1 and stage #2: the nonsecure application could try to map the secure buffer to check the data have the expected value. but that quite unsafe (and will trig a security violation on a real hardened setup). So it's debug stuff and must be optional.